### PR TITLE
Adds http bytes request(tested in hashlink)

### LINKED ISSUE #2586: haxe.Http class should not use Strings to represent binary data
We have modified the C++ String implementation to correctly support UTF-8 strings.  Our implementation does not accept improperly encoded UTF-8 strings, and we have found that when using the haxe.Http class, the representation of binary data as Strings in functions like onData and setPostData causes problems.

Using a String as a container for binary data implies that the String internal representation is in bytes; but this is a problematic assumption to make when the String internal representation is UTF-8.

Another issue is the fact that binary data used in onData and setPostData may have embedded 0 values (nulls), and embedded nulls are in general at odds with string representations.

Why don't onData and setPostData use Bytes objects instead of Strings?

### PR DESCRIPTION
**Pull request by mundusnine**
- Response data is set to be of type Any
- When requesting data that has the "Accept-Ranges" set to bytes in the responseHeaders, we set the responseData to be bytes.
- Tested with zip/image files on the hashlink platform.
- Could be added to other platforms using the same logic.
- Partially fixes [this](https://github.com/HaxeFoundation/haxe/issues/2586) issue.

### CODE REVIEW

File: std/sys/Http.hx [hunk 1]
```diff
@@ -58,12 +58,20 @@ class Http extends haxe.http.HttpBase {
 			onError(e);
 		}
 		customRequest(post,output);
-		if(!err)
+		if(!err){
 		#if neko
 			onData(responseData = neko.Lib.stringReference(output.getBytes()));
 		#else
-			onData(responseData = output.getBytes().toString());
+			var respType = responseHeaders.get("Accept-Ranges");
```
Comment (haxiomic) [inline std/sys/Http.hx]: I'm not sure checking Accept-Ranges is the right approach to determine if the response is binary – this just implies the resource can be accessed with range requests (which many binary resources will not set).

For the most predictable behavior we need to mirror whatever the industry standard is as closely as possible. Checking the "Content-Type" header would be the normal approach (details here: https://tools.ietf.org/html/rfc7231#section-3.1.1.5), however if this is not set we need to choose a sensible guess. The specification says
> If a Content-Type header field is not present, the recipient
   MAY either assume a media type of "application/octet-stream"
   ([RFC2046], Section 4.5.1) or examine the data to determine its type.

But I'd want to review what popular implementations are actually doing (like web browsers) so we're conforming to everyone's expectations.

**Comment by haxiomic** [comment]
I agree adding support for binary data in http requests is an important change to make, but I think we've also got to be very careful about getting it right. Changing HttpBase.hx will affect _a lot_ of projects, so if we have to make a breaking change we ideally want to make it future proof.

As the response is usually a `String`, maybe it makes sense to always return an object that unifies with `String` but lets you access the binary data easily? This way existing projects will still work as normal but you can access binary responses when you need them. (In JavaScript you [override the responseType to do this](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Sending_and_Receiving_Binary_Data#Receiving_binary_data_using_JavaScript_typed_arrays)).

I'd say this is a good starting point but we need to do a design pass + a set of tests so we can be confident we've got it right.

**Comment by mundusnine** [comment]
@haxiomic  What people proposed in [this issue](https://github.com/HaxeFoundation/haxe/issues/2586) is to add a onBytes() function definition in the HttpBase.hx and always send the Bytes since we always get them anyways. This would let the user define the function if he needs to get Bytes.
I could do it that way if you think it would minimize the impact on other projects. Essentially people who already use onData would not be affected but everybody could define the new onBytes function like is done to the onData func.

**Comment by haxiomic** [comment]
@mundusnine go for it :), we'll have to see what the others feel but that approach leaves existing code untouched and moves the situation forward

I think with having both `onData` and `onBytes` isn't a clear API unfortunately (it's not clear from the names why you'd have both...), so maybe we deprecate `onData` and rename it to `onString` (or aim to remove it altogether in favor of having the user do `bytes.getString()`)

File: std/sys/Http.hx [hunk 2]
```diff
@@ -58,12 +58,14 @@ class Http extends haxe.http.HttpBase {
 			onError(e);
 		}
 		customRequest(post,output);
-		if(!err)
+		if(!err){
+			onBytes(output.getBytes());
```
Comment (haxiomic) [inline std/sys/Http.hx:62]: I think we can't call `getBytes()` here twice (see docs on https://api.haxe.org/haxe/io/BytesOutput.html#getBytes)

Best to store it in a variable before passing to onBytes and onData

Comment (haxiomic) [reply std/sys/Http.hx:62]: (the reason being is that this works on some platforms but fails on others)

**Pull request closed without merging**